File: Thread-Benchmark/2-AsyncTask/Case-9/MainActivity.java
package com.test.anr.backgroundThread;

import android.app.Activity;
import android.os.AsyncTask;
import android.os.Bundle;

import java.io.FileWriter;
import java.io.IOException;

public class MainActivity extends Activity {
    FileWriter fw;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        try {
            fw = new FileWriter("s.txt");
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        runCase();
        runCase();
    }

    private void runCase() {
        AsyncTask<String, Void, Void> t = new AsyncTask<String, Void, Void>() {
            @Override
            protected Void doInBackground(String... strings) {
                String label = (strings != null && strings.length > 0) ? strings[0] : "case09-no-param";
                IOBlockingCall("case09-string-param-" + label);
                return null;
            }
        };
        t.execute("tag09");
    }

    private void IOBlockingCall(String tag) {
        try {
            if (fw != null) {
                fw.write("IOBlockingCall: " + tag + "\n");
                fw.flush();
            }
        } catch (IOException e) {
        }
    }
}

File: Thread-Benchmark/6-coroutines/Case-1/MainActivity.kt
package com.test.anr.backgroundThread

import android.app.Activity
import android.os.Bundle
import java.io.FileWriter
import java.io.IOException
import kotlinx.coroutines.*

class MainActivity : Activity() {

    private var fw: FileWriter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        try {
            fw = FileWriter("s.txt")
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        runCase()
        runCase()
    }

    private fun runCase() {
        GlobalScope.launch(Dispatchers.IO) {
            IOBlockingCall("case01-global-io")
        }
    }

    private fun IOBlockingCall(tag: String) {
        try {
            fw?.write("IOBlockingCall: $tag\n")
            fw?.flush()
        } catch (e: IOException) {
        }
    }
}
